sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/download/http/HttpDownloadFileTask.java

```diff
@@ -0,0 +1,196 @@
+        try {
+            File directory = mTargetFile.getParentFile();
+            if (directory == null || !(directory.exists() || directory.mkdirs())) {
+                throw new IOException("Could not create the directory for file:" + mTargetFile.getAbsolutePath());
+            }
+            if (mTargetFile.exists()) {
+
+                //noinspection ResultOfMethodCallIgnored
+                mTargetFile.delete();
+            }
+
+            /* Create connection */
+            URL url = new URL(mDownloadUri.toString());
+            TrafficStats.setThreadStatsTag(THREAD_STATS_TAG);
+            URLConnection connection = createConnection(url, MAX_REDIRECTS);
+            connection.connect();
+            String contentType = connection.getContentType();
+            if (contentType != null && contentType.contains("text")) {
+
+                /* This is not the expected APK file. Maybe the redirect could not be resolved. */
+                throw new IOException("The requested download does not appear to be a file.");
+            }
+
+            /* Download the release file. */
+            long totalBytesDownloaded = downloadFile(connection.getInputStream(), connection.getContentLength());
+            if (totalBytesDownloaded > 0) {
+                HttpConnectionReleaseDownloader downloader = mDownloader.get();
+                if (downloader != null) {
+                    downloader.onDownloadComplete(mTargetFile);
+                }
+            }
+        } catch (IOException e) {
+            HttpConnectionReleaseDownloader downloader = mDownloader.get();
+            if (downloader != null) {
+                downloader.onDownloadError(e.getMessage());
+            }
+        } finally {
+            TrafficStats.clearThreadStatsTag();
+        }
+        return null;
+    }
+
+    /**
+     * Performs IO operation to download .apk file through HttpConnection.
+     * Saves .apk file to the mApkFilePath.
+     *
+     * @param inputStream  TODO
+     * @param lengthOfFile TODO
+     * @return total number of downloaded bytes.
+     * @throws IOException if connection fails
+     */
+    @SuppressWarnings("TryFinallyCanBeTryWithResources")
+    private long downloadFile(InputStream inputStream, long lengthOfFile) throws IOException {
+        InputStream input = null;
+        OutputStream output = null;
+        try {
+            input = new BufferedInputStream(inputStream);
+            output = new FileOutputStream(mTargetFile);
+            byte[] data = new byte[WRITE_BUFFER_SIZE];
+            int count;
+            long totalBytesDownloaded = 0, reported = 0;
+            while ((count = input.read(data)) != -1) {
+                totalBytesDownloaded += count;
+                output.write(data, 0, count);
+
+                /* Update the progress. */
+                if (totalBytesDownloaded >= reported + UPDATE_PROGRESS_BYTES_COUNT || totalBytesDownloaded == lengthOfFile) {
```

Comment:
I understand that this should not be the case, but what if there's some internal glitch and we hang forever..?
```suggestion
                if (totalBytesDownloaded >= reported + UPDATE_PROGRESS_BYTES_COUNT || totalBytesDownloaded >= lengthOfFile) {
```

Reply:
No, it's just progress update. Let's keep as is